lib/Controller/DevicesApiController.php

```diff
@@ -113,6 +113,10 @@ public function getDevicePoints($id, $pruneBefore=0) {
      */
     public function addDevicePoint($apiversion, $lat, $lng, $timestamp=null, $user_agent=null, $altitude=null, $battery=null, $accuracy=null) {
```

Comment:
The same method is also in `DevicesController`, so maybe fix that as well?

Reply:
Also, I think it's generally a better idea to use parameter types here instead of checking whether a type is numeric or not, for example:

```suggestion
    public function addDevicePoint(int $apiversion, ?float $lat, ?float $lng, int $timestamp=null, string $user_agent=null, float $altitude=null, float $battery=null, float $accuracy=null) {
```

That way it limits the types possible when this method is called. It might also be a good idea to use `declare(strict_types=1);` at the start of the file. However, this might break things for other methods, so it needs to be tested accordingly.

As a side note: There is also [`@param`](https://docs.nextcloud.com/server/17/developer_manual/app/requests/controllers.html#casting-parameters) for use in documentation comments. In theory this should be inferred by type annotations as well, even when using strict types, but I haven't verified whether this really is the case.

Reply:
is adding the types enough or do we still need the normalization function? Also, what's the meaning of `?` in `?float`? I couldn't find it on google :/

Reply:
Also, I don't understand the difference between `DevicesApiController` and `DevicesController`?

They are _almost_ the same but not. I don't feel comfortable enough with the codebase to try and extract proper abstractions for avoid code duplication.

I can add the same normalization step in `DevicesController` as a quick win, if you prefer it

Reply:
The `?typename` means that it is nullable, see https://wiki.php.net/rfc/nullable_types.

Reply:
As for `DevicesController` versus `DevicesApiController`: I don't know the difference either, but just did a quick grep for the method name in question and found it in both. A diff between the two seems that the methods that occur in both are only different by the usage of `@NoCSRFRequired` and `@CORS`, so I think that definitely smells like it could be deduplicated, certainly the `addDevicePoint` function, since their method bodies are the same.

Reply:
I just explained the difference in the latest comment on this thread.

The code could be factorized indeed but I think we should keep the two distinct controllers to make sure one is getting the CRSF check and the other does not.